File: apps/api/src/fator_r/core/cnpj.py
"""CNPJ: normalização, validação dos dígitos verificadores e formatação."""

import re

_PESOS_DV1 = (5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2)
_PESOS_DV2 = (6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2)


class CnpjInvalido(ValueError):
    pass


def _digito(numeros: str, pesos: tuple[int, ...]) -> str:
    resto = sum(int(n) * p for n, p in zip(numeros, pesos, strict=True)) % 11
    return "0" if resto < 2 else str(11 - resto)


def normalizar_cnpj(valor: str) -> str:
    """Devolve só os 14 dígitos de um CNPJ válido; levanta CnpjInvalido caso contrário."""
    digitos = re.sub(r"\D", "", valor)
    if len(digitos) != 14 or len(set(digitos)) == 1:
        raise CnpjInvalido("CNPJ inválido")
    dv1 = _digito(digitos[:12], _PESOS_DV1)
    dv2 = _digito(digitos[:12] + dv1, _PESOS_DV2)
    if digitos[12:] != dv1 + dv2:
        raise CnpjInvalido("CNPJ inválido")
    return digitos


def cnpj_valido(valor: str) -> bool:
    try:
        normalizar_cnpj(valor)
    except CnpjInvalido:
        return False
    return True


def formatar_cnpj(digitos: str) -> str:
    d = normalizar_cnpj(digitos)
    return f"{d[:2]}.{d[2:5]}.{d[5:8]}/{d[8:12]}-{d[12:]}"

File: apps/web/lib/cnpj.ts
/** Máscara e checagem de CNPJ para feedback imediato. A api valida de novo. */

export function mascararCnpj(valor: string): string {
  const d = valor.replace(/\D/g, "").slice(0, 14);
  return d
    .replace(/^(\d{2})(\d)/, "$1.$2")
    .replace(/^(\d{2})\.(\d{3})(\d)/, "$1.$2.$3")
    .replace(/\.(\d{3})(\d)/, ".$1/$2")
    .replace(/(\d{4})(\d)/, "$1-$2");
}

function digito(numeros: string, pesos: number[]): string {
  const resto = numeros.split("").reduce((s, n, i) => s + Number(n) * pesos[i], 0) % 11;
  return resto < 2 ? "0" : String(11 - resto);
}

export function cnpjValido(valor: string): boolean {
  const d = valor.replace(/\D/g, "");
  if (d.length !== 14 || new Set(d).size === 1) return false;
  const dv1 = digito(d.slice(0, 12), [5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]);
  const dv2 = digito(d.slice(0, 12) + dv1, [6, 5, 4, 3, 2, 9, 8, 7, 6, 5, 4, 3, 2]);
  return d.slice(12) === dv1 + dv2;
}
